### ISSUE TITLE
Объекты и интерфейсы

### ISSUE THREAD

**Issue by acrm** [Rideress/JavaLearning#3]
В прошлой задаче мы использовали классы для объединения статических данных и методов, которые их используют.  Такие классы подходят для описания чего-то уникального, что существует только в одном экземпляре в рамках нашей программы. С помощью иерархии вложенности мы можем отразить связь между такими классами. Например, в модели компьютера мы вкладывали классы отдельных устройств в класс подсистемы, чтобы показать, что эти устройства могут находится в рамках этой подсистемы, но не в других подсистемах. И каждую систему и каждое устройство мы рассматривали как уникальную сущность, которую можно описать с помощью статических членов соответствующего класса. Но вообще говоря, это было определённое упрощение. Например, допустим что в нашем компьютере присутствует два монитора. Каждый из них имеет своё отдельное название и включается/выключается независимо от другого. Как нам отразить это на модели? Конечно, мы могли бы включить описание двух этим мониторов в один класс:
```java
class OutputSystem {
    static class Display {
        static String name1 = "Монитор 1";
        static String name2 = "Монитор 2";

        static void turnOnDisplay1() {
           System.out.println(name1 + " включён"); 
       }
      
        static void turnOnDisplay2() {
           System.out.println(name2 + " включён"); 
       }
    }
}
```
Но это уже не правильное использование классов. Класс должен содержать описание какой-то одной сущности. Если таких сущностей несколько, то их надо описывать разными классами. Но если при этом по своему содержанию эти сущности одинаковы - содержат одинаковый набор полей и методов - то нет необходимости на каждую описывать отдельный класс, можно просто создавать разные экземпляры одного и того же класса. Просто у одного объекта будут одни значения полей, а у другого - другие. Так же не стоит описывать несколько идентичных классов, различающихся только именами, например:
```java
class OutputSystem {
    static class Display1 {
        static String name = "Монитор 1";

        static void turnOn() {
           System.out.println(name + " включён"); 
       }
    }
    static class Display2 {
        static String name = "Монитор 2";

        static void turnOn() {
           System.out.println(name + " включён"); 
       }
    }
}
```
Вместо таких классов опять же правильно написать один общий класс и создавать его экземпляры с разными значениями полей.

Поля можно задавать двумя способами: либо в конструкторе, либо через сеттеры.
- Конструктор - это метод, который вызывается автоматически при создании экземпляра класса. Он может принимать параметры и на основе этих параметров может инициализировать поля класса.
- Сеттер - это обычный метод, который называется set*Field*(value), и задаёт переданным ему параметром значение того поля, для которого он создан

**Comment by Rideress** [comment]
Получается ключевое слово .this, при создании объекта и вызове метода, будет создавать поля для каждого объекта свои и у всех объектов при разных значениях этих полей будут разные ссылки?

**Comment by acrm** [comment]
Неправильно здесь говорить, что слово ```this``` "создаёт поля". Если уж на то пошло, то выделяет память для объекта и создаёт его поля слово ```new```, а в переменную ```this``` просто сохраняется ссылка на этот объект. И да, у разных объектов эти ссылки будут указывать на разные области памяти, и поля, которые содержаться в этих объектах, тоже у разных объектов располагаются в разных адресах.

P.S. Обрати внимание, добавил подзадачу задачу со звёздочкой, более сложную

**Comment by Rideress** [comment]
А в задаче получается системы будут представляться классами, а устройства объектами этих классов, я правильно понял?

**Comment by acrm** [comment]
Нет, все будут объектами, начиная от компьютера. Объектами соответствующих классов: компьютеров, подсистем ввода/вывода, устройств чтения/записи, процесооров и т.д.

**Comment by acrm** [comment]
```this``` используется так:
```java
class A {
    public void methodA1() {
        this.methodA2();    // this хранит ссылку на текущий объект, который можно назвать a1
                                      // обращаемся к методу methodA2() объекта a1
    }
    public void methodA2() {
        B b1 = new B();    // b1 хранит ссылку на объект b1
        b1.methodB1();    // обращаемся к методу methodB1() объекта b1
    }
}
class B {
    public void methodB1() {
        this.methodB2();    // this тоже хранит ссылку на объект b1, когда мы находимся внутри этого объекта
                                      // обращаемся к методу methodB2() объекта b1
    }
    public void methodB2() {
    }
}
```

**Comment by Rideress** [comment]
получается когда мы используем просто this то у нас создается объект?

**Comment by acrm** [comment]
Нет, ещё раз: ```this``` — это просто переменная. Когда мы обращаемся к переменной, у нас от этого не создаются нигде никакие объекты. Переменная просто хранит ссылку на уже созданный объект. В случае ```this``` — это всегда ссылка на текущий объект. Создаётся объект, как я уже писал, ключевым словом ```new```.

Представь, что ты находишься в тёмной комнате и не видишь ни себя самого, ни вещей, которые находятся в этой комнате. Но, допустим, ко всем предметам в этой комнате одними концами привязаны длинные прочные нити, а другие концы этих нитей находятся у тебя в руке, причём с бирками, на которых азбукой Брайля написано, какой предмет привязан к этой нити. Допустим, ты без проблем читаешь азбуку Брайля и можешь по бирке определить, к какому предмету ведёт та или иная нить. Захотел сесть на диван — нащупал в своей руке бирку с надписью "couch", прошёл вдоль нити, идущей от неё, дошёл до конца нити, нащупал края дивана, сел. Так вот, в этой аналогии предметы в тёмной комнате — это объекты в памяти, нити — это ссылки на них, а бирки — это переменные, в которых ссылки хранятся. Продолжим нашу фантазию: представим, что у нас не одна такая комната, а целый длинный коридор, вдоль которого находятся двери во множество таких тёмных комнат. И чтобы не заблудится в этом коридоре к двери в каждую комнату тоже привязана нитка с биркой, которую мы держим в руке. А внутри каждой комнаты у входа есть доска, на которой закреплены бирки с нитями на все предметы в этой комнате. Находясь у входа можно нащупать нужную бирку и от неё пройти нужного предмета. при этом бирка снимается с доски и назад по этой ниточке мы не сможем вернуться назад к доске, потому что теперь один конец нити у нас в руке, а второй у предмета, к которому мы идём. Поэтому, чтобы иметь возможность вернуться назад к к доске с другими бирками и выходу из комнаты, мы снимаем с доски ещё одну бирку с надпись "this" — эта бирка связана с нитью, которая привязана к самой доске. С помощью неё ты всегда можешь вернуться к тому месту, где закреплены бирки на все предметы в этой комнате. 

В общем, это примерная аналогии, если  углубляться в детали, то придётся всё больше и больше ограничений и допущений добавлять, чтобы это было похоже на то, что в действительности происходит в программе. В программе для того, чтобы получить доступ к какому-то объекту или переменной, всегда нужно знать её точный адрес в памяти — это аналог ниточки, которая ведёт к предмету. This - это адрес текущего объекта в глобальной памяти, а все поля этого объекта имеют относительный адрес, связанный с адресом своего объекта. Адрес объекта — это адрес его первого поля. Например, пусть объект типа Point состоит из двух полей: X и Y. Допустим, каждое поле занимает в памяти 4 байта. Значит адрес поля X будет равен адресу объекта, а адрес поля Y равен адресу объекта + 4 байта. Тогда если объект point будет размещён в памяти по адресу 700, то адрес поля X будет тоже 700, а поля Y — 704. Суть в том, что нам как правило при написании программы не нужно точно знать адреса, в которых находятся объекты и их поля (да это, обычно и не возможно: при каждом запуске программы её объектны располагаются по новым адресам). Нам важно знать, что эти адреса сохраняются в переменные с понятными нам именами и мы можем пользоваться этими переменными, чтобы обращаться к объектам и полям. Запись ```point.Y``` для компьютера, выполняющего нашу программу, означает «возьми адрес, хранящийся в переменной point, добавь к нему смещение, которое соответствует полю Y, и прочитай 4 байта, которые записаны, начиная с этого адреса».

Давай рассмотрим ещё один пример, возвращаясь от аналогий с ниточками и внутренностей адресной арифметики к более привычному нам коду на Java:
```java
new A(); // Здесь мы создали новый объект типа A, но никуда не сохранили ссылку на него
A a;    // Здесь мы объявили переменную, в которую может быть сохранена ссылка на объект типа A. Но пока ещё она пустая.
a = new A(); // А вот тут мы создали новый объект (не тот же самый, что и двумя строчками выше) и ссылку на него сохранили в переменную a
```

**Comment by acrm** [comment]
Кстати, не совсем понял, что тебя смущает в конкатенации объектов

**Comment by Rideress** [comment]
У меня есть вопрос, я до сих пор не понимаю как мне реализовать эту задачу, примерная идея такая:

1. Все классы (Computer, InputDevices, OutputDevices, SystemUnit) нужно сделать не статическими, так как у статических методов мы не можем использовать  this.
2. Метод включения либо придется описывать для каждого класса свой, либо надо сделать так чтобы он был один для всех классов, но брал значение из объекта, но тогда надо чтобы объекты хранили ссылку на название, но так как объекты являются экземплярами своих классов это не получится.
3. Использовать ту же реализацию что и раньше, но только заменить классы устройств на объекты.

**Comment by acrm** [comment]
1. Да, совершенно верно, в статических методах мы не можем использовать ```this``` (Но можем использовать имя класса для обращения к собственным статическим членам). Но, как я уже говорил, в Java классы сами  по себе не являются статичными, любой класс может содержать и статические и нестатические члены. То что наши вложенные классы обозначены как статические означает то, что доступ к этим классам осуществляется через имя класса верхнего уровня (а не через имя объекта класса верхнего уровня, если бы вложенный класс не был бы объявлен статично). Но в любом случае это не мешает использовать в динамичных методах этого класса использовать ```this```.
2. Всё верно, хотя не понял в чём проблема хранить ссылку на название и обращаться к этой ссылке из методов в экземплярах класса? В любом случае, давай пока сделаем для каждого класса свой метод включения (сейчас, вроде бы, так и сделано?)
3. В целом структура компьютера должна сохраняться, но теперь она будет передаваться не вложенностью классов (её можно и оставить, а можно и развернуть, сделав все классы верхнего уровня. Компактнее будет оставить их вложенными), а через иерархии ссылок: объект компьютера хранит ссылки на объекты подсистем, объекты подсистем хранят ссылки на объекты устройств и т.д.

Я рекомендую тебе попробовать уже что-нибудь написать и залить, чтобы я мог комментировать и уточнять прямо по коду.
